File: Binary_Tree_TopView/Binary_Tree_TopView.cpp
#include <iostream>
#include <queue>
#include <map>
using namespace std;

// Structure of binary tree leftSide , head , rightSide
struct Node
{
    Node *leftSide;
    Node *rightSide;
    int head;
    int data;
};

// function to create a new node
Node *newNode(int key)
{
    Node *node = new Node();
    node->leftSide = node->rightSide = NULL;
    node->data = key;
    return node;
}

// function should print the topView of
// the binary tree
void topview(Node *root)
{
    if (root == NULL)
        return;
    queue<Node *> q;
    map<int, int> m;
    int head = 0;
    root->head = head;

    // push node and horizontal distance to queue
    q.push(root);

    cout << "The top view of the tree is : \n";

    while (q.size())
    {
        head = root->head;

        // count function returns 1 if the container
        // contains an element whose key is equivalent
        // to head, or returns zero otherwise.
        if (m.count(head) == 0)
            m[head] = root->data;
        if (root->leftSide)
        {
            root->leftSide->head = head - 1;
            q.push(root->leftSide);
        }
        if (root->right)
        {
            root->rightSide->head = head + 1;
            q.push(root->rightSide);
        }
        q.pop();
        root = q.front();
    }

    for (auto i = m.begin(); i != m.end(); i++)
    {
        cout << i->second << " ";
    }
}

// Driver Program to test above functions
int main()
{
    Node *root = newNode(1);
    root->leftSide = newNode(2);
    root->rightSide = newNode(3);
    root->leftSide->right = newNode(4);
    root->leftSide->rightSide->rightSide = newNode(5);
    root->leftSide->rightSide->rightSide->rightSide = newNode(6);
    cout << "Following are nodes in top view of Binary Tree\n";
    topview(root);
    return 0;
}

File: Binary_Tree_Top_View/Binary_Tree_Top_View.java
import java.util.Queue;
import java.util.TreeMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Map.Entry;

// class to create a node 
class Node {
    int data;
    Node left, right;

    public Node(int data) {
        this.data = data;
        left = right = null;
    }
}

// class of binary tree
class BinaryTree {
    Node root;

    public BinaryTree() {
        root = null;
    }

    // function should print the topView of
    // the binary tree
    private void TopView(Node root) {
        class QueueObj {
            Node node;
            int hd;

            QueueObj(Node node, int hd) {
                this.node = node;
                this.hd = hd;
            }
        }
        Queue<QueueObj> q = new LinkedList<QueueObj>();
        Map<Integer, Node> topViewMap = new TreeMap<Integer, Node>();

        if (root == null) {
            return;
        } else {
            q.add(new QueueObj(root, 0));
        }

        System.out.println("The top view of the tree is : ");

        // count function returns 1 if the container
        // contains an element whose key is equivalent
        // to hd, or returns zero otherwise.
        while (!q.isEmpty()) {
            QueueObj tmpNode = q.poll();
            if (!topViewMap.containsKey(tmpNode.hd)) {
                topViewMap.put(tmpNode.hd, tmpNode.node);
            }

            if (tmpNode.node.left != null) {
                q.add(new QueueObj(tmpNode.node.left, tmpNode.hd - 1));
            }
            if (tmpNode.node.right != null) {
                q.add(new QueueObj(tmpNode.node.right, tmpNode.hd + 1));
            }

        }
        for (Entry<Integer, Node> entry : topViewMap.entrySet()) {
            System.out.print(entry.getValue().data);
        }
    }

    // Driver Program to test above functions
    public static void main(String[] args) {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.right = new Node(4);
        tree.root.left.right.right = new Node(5);
        tree.root.left.right.right.right = new Node(6);
        System.out.println("Following are nodes in top view of Binary Tree");
        tree.TopView(tree.root);
    }

}
